internal/provider/virtual_cluster_credentials_resource.go

```diff
@@ -0,0 +1,257 @@
+package provider
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/hashicorp/terraform-plugin-framework/resource"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/planmodifier"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/stringplanmodifier"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+	"github.com/warpstreamlabs/terraform-provider-warpstream/internal/provider/api"
+)
+
+// Ensure the implementation satisfies the expected interfaces.
+var (
+	_ resource.Resource              = &virtualClusterCredentialsResource{}
+	_ resource.ResourceWithConfigure = &virtualClusterCredentialsResource{}
+)
+
+// NewVirtualClusterCredentialsResource is a helper function to simplify the provider implementation.
+func NewVirtualClusterCredentialsResource() resource.Resource {
+	return &virtualClusterCredentialsResource{}
+}
+
+// virtualClusterCredentialsResource is the resource implementation.
+type virtualClusterCredentialsResource struct {
+	client *api.Client
+}
+
+// virtualClusterCredentialsModel maps credentials schema data.
+type virtualClusterCredentialsModel struct {
+	ID               types.String `tfsdk:"id"`
+	Name             types.String `tfsdk:"name"`
+	UserName         types.String `tfsdk:"username"`
+	Password         types.String `tfsdk:"password"`
```

Comment:
This is the first time I look at the actual implementation of a terraform provider, so maybe it's a stupid question, but does this mean that we are storing the password as part of the tf state?

Reply:
You are correct, the password is stored in the tf state. If we don't save it, the password would be lost.

At line 110 we declare it as a `sensitive` attribute so terraform will do its best to never display it.
It is not uncommon for terraform state to store sensitive values and the user should use a secured backend in that case.
See https://developer.hashicorp.com/terraform/language/state/sensitive-data.

Reply:
ok thanks for the pointer 👍